webhook/client.go

```diff
@@ -0,0 +1,99 @@
+package webhook
+
+import (
+	"crypto/hmac"
+	"crypto/sha256"
+	"encoding/hex"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/stripe/stripe-go"
+)
+
+const ToleranceDefault int64 = 300
+const ToleranceIgnoreTimestamp int64 = 0
+
+// ConstructEvent initializes an Event object from a JSON payload.
+// It returns an non-nil error when the payload is not valid JSON or when signature verification fails.
+// payload is the webhook request body, i.e. `ioutil.ReadAll(r.Body)`.
+// sigHeader is the webhook Stripe-Signature header, i.e. `r.Header.Get("Stripe-Signature")`.
+// secret is your Signing Secret, i.e. `"whsec_XYZ"`.  See https://dashboard.stripe.com/webhooks
+// tolerance (suggested 300) is the max difference in seconds between now and Stripe-Signature's timestamp. If the difference is greater than this tolerance, the signature is rejected and a non-nil error is returned.  If tolerance is 0 or less, then the timestamp is not checked.
+// NOTE: your requests will only have Stripe-Signature if you have clicked to reveal your secret
+func ConstructEvent(payload []byte, sigHeader string, secret string, tolerance int64) (stripe.Event, error) {
+	e := stripe.Event{}
+
+	if err := json.Unmarshal(payload, &e); err != nil {
+		return e, fmt.Errorf("Failed to parse webhook body json: %s", err.Error())
+	}
+
+	if sigHeader == "" {
+		return e, errors.New("Missing Stripe-Signature header")
+	}
+
+	// sigHeader looks like "t=1495999758,v1=ABC,v1=DEF,v0=GHI"
+
+	// First extract the timestamp
+	var t string
+	pairs := strings.Split(sigHeader, ",")
+	for _, pair := range pairs {
+		parts := strings.Split(pair, "=")
```

Comment:
So this is what we want: an `a=b=c` entry should be decoded as a pair (prefix) `a` / (value) `b=c`.

That said, you're correct that we should also verify that there _are_ two elements and not just one. An entry without any `=` character would be equally invalid and the code would fail when trying to access `parts[1]`.